### PR TITLE
VUU21: Implement the screenshot functionality into the Save Layout component

### PR DESCRIPTION
**Pull request by cfisher-scottlogic**
## Scope

[VUU21](https://github.com/ScottLogic/finos-vuu/issues/21)

This incorporates the work done in the screenshot spike (#19) by adding screenshot functionality into the save dialog component.

## Implementation

Screenshot functionality is provided with the [html-to-image](https://www.npmjs.com/package/html-to-image) package in `screenshot-utils.ts`. The screenshot is currently stored in browser (local) storage as a base64 encoded string. Where fonts are imported externally via a link, the `crossorigin="anonymous"` tag has been added due to CORS issues, see [this stackoverflow post](https://stackoverflow.com/questions/49993633/uncaught-domexception-failed-to-read-the-cssrules-property)  and [this html-to-image GitHub issue](https://github.com/bubkoo/html-to-image/issues/40) discussing the issue.

A unit test has been written for `takeScreenshot()`. The html-to-image package had to be mocked out as the browser operations it relies on are not available in the test environment (neither happy-dom nor jsdom).

A Cypress e2e test has been written to check a screenshot can be taken and rendered successfully from the UI. I've restructured the directory by separating the `./support` files relevant to component & e2e tests.

## Screenshots

image

### CODE REVIEW

File: vuu-ui/.eslintrc.json
Comment (cfisher-scottlogic) [inline vuu-ui/.eslintrc.json:1]: Looks like I've got some pretty aggressive auto-formatting happening as there are only a couple of relevant changes in this file. I'll change my formatting settings and revert the file to tidy up the PR.

File: vuu-ui/cypress/e2e/layout-management/screenshot.cy.js
Comment (cfisher-scottlogic) [inline vuu-ui/cypress/e2e/layout-management/screenshot.cy.js:27]: I'm going to explore alternative selector options to hopefully land on something which is more indicative of what a user would actually do thus more accessible.

File: vuu-ui/cypress/tests/checkAccessibility.tsx
Comment (cfisher-scottlogic) [inline vuu-ui/cypress/tests/checkAccessibility.tsx:1]: Unnecessary line commenting. I'll revert this back.

File: vuu-ui/packages/vuu-shell/src/layout-management/SaveLayoutPanel.css
Comment (cfisher-scottlogic) [inline vuu-ui/packages/vuu-shell/src/layout-management/SaveLayoutPanel.css:1]: WIP from #17, not relevant to this PR.

File: vuu-ui/showcase/src/examples/Apps/NewTheme.examples.tsx
Comment (cfisher-scottlogic) [inline vuu-ui/showcase/src/examples/Apps/NewTheme.examples.tsx:46]: I'm going to change the selector for the node passed in to something more reliable.

File: vuu-ui/showcase/src/examples/Layout/SavePanel.examples.tsx
Comment (cfisher-scottlogic) [inline vuu-ui/showcase/src/examples/Layout/SavePanel.examples.tsx:1]: WIP from #17, not relevant to this PR.

File: vuu-ui/showcase/src/examples/Layout/index.ts
Comment (cfisher-scottlogic) [inline vuu-ui/showcase/src/examples/Layout/index.ts:1]: WIP from #17, not relevant to this PR.

**Comment by cfisher-scottlogic** [comment]
We need to decide what the base of this PR should be.

As this is adding functionality to [VUU-17-layoutSave](https://github.com/ScottLogic/finos-vuu/tree/feature/VUU-17-layoutSave) it makes sense to make that the base branch. 

But, if this (VUU21) is merged into that branch (VUU17) before VUU17 is merged into the feature branch, it could really bloat the PR for VUU17 and make it harder to review non-screenshot related changes. 

So we can either wait till VUU17 is merged into the feature branch before merging this PR in. Or we can merge this PR straight into the feature branch.

**Comment by cfisher-scottlogic** [comment]
> We need to decide what the base of this PR should be.
> 
> As this is adding functionality to [VUU-17-layoutSave](https://github.com/ScottLogic/finos-vuu/tree/feature/VUU-17-layoutSave) it makes sense to make that the base branch.
> 
> But, if this (VUU21) is merged into that branch (VUU17) before VUU17 is merged into the feature branch, it could really bloat the PR for VUU17 and make it harder to review non-screenshot related changes.
> 
> So we can either wait till VUU17 is merged into the feature branch before merging this PR in. Or we can merge this PR straight into the feature branch.

Discussed offline. Base branch will be [VUU-17-layoutSave](https://github.com/ScottLogic/finos-vuu/tree/feature/VUU-17-layoutSave) for reviewing but should be merged into the feature branch when #17 has been merged.

File: vuu-ui/cypress/e2e/layout-management/screenshot.cy.js
Comment (cfisher-scottlogic) [reply vuu-ui/cypress/e2e/layout-management/screenshot.cy.js:27]: We're wrapping up improvements to the e2e testing strategy (selectors, POM, etc) into a separate ticket as it'll be more of an ongoing effort. I'll give the TODOs a ticket number to track this.

File: vuu-ui/packages/vuu-ui-controls/src/list/CheckboxIcon.css
Comment (cfisher-scottlogic) [inline vuu-ui/packages/vuu-ui-controls/src/list/CheckboxIcon.css:1]: Again, aggressive formatting...

File: vuu-ui/showcase/templates/index.html
Comment (pling-scottlogic) [inline vuu-ui/showcase/templates/index.html:1]: Same as above re. indentation.

**Review by pling-scottlogic — COMMENTED** [review]
Just a couple of minor points about indentation. I'll leave it to someone with more JS/TS experience for approval.

File: vuu-ui/showcase/src/examples/Apps/NewTheme.examples.tsx
Comment (heswell) [inline vuu-ui/showcase/src/examples/Apps/NewTheme.examples.tsx:83]: I've pushed a small change to main. In this MenuActionClosePopup 'action', you have an options object, populated from the Tab at the time the ContextMernu is created. Previoulsy, it just contained the index of the tab from which the popupMenu was launched. Now it also contains the controlledComponentId - the id of the associated TabPanel (its also visible in the html as the aria attribute 'aria-controls'.

Comment (cfisher-scottlogic) [reply vuu-ui/showcase/src/examples/Apps/NewTheme.examples.tsx:88]: Brill thanks Steve. 

[aa898c2](https://github.com/ScottLogic/finos-vuu/pull/23/commits/aa898c20d1ff7f17315d89bf0996fa47f13c5518)

File: vuu-ui/showcase/src/examples/Apps/NewTheme.examples.tsx
Comment (cfisher-scottlogic) [reply vuu-ui/showcase/src/examples/Apps/NewTheme.examples.tsx]: [aa898c2](https://github.com/ScottLogic/finos-vuu/pull/23/commits/aa898c20d1ff7f17315d89bf0996fa47f13c5518)

File: vuu-ui/showcase/templates/index-preview.html
Comment (cfisher-scottlogic) [reply vuu-ui/showcase/templates/index-preview.html:1]: [0c94ea8](https://github.com/ScottLogic/finos-vuu/pull/23/commits/0c94ea8cd33bce876ee0d55ab7191a400c0c05a3)

File: vuu-ui/showcase/templates/index.html
Comment (cfisher-scottlogic) [reply vuu-ui/showcase/templates/index.html:1]: [0c94ea8](https://github.com/ScottLogic/finos-vuu/pull/23/commits/0c94ea8cd33bce876ee0d55ab7191a400c0c05a3)

**Review by vferraro-scottlogic — APPROVED** [review]

**Pull request merged**
